compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+/// we replace either all or none of the occurrences of `*_1`.
+///
+/// Some care has to be taken when `_1` is copied in other locals.
+///   _1 = &raw? mut? _2;
+///   _3 = *_1;
+///   _4 = _1
+///   _5 = *_4
+/// In such cases, fully substituting `_1` means fully subtituting all of the copies.
+///
+/// For immutable borrows, we do not need to preserve such uniqueness property,
+/// so we perform all the possible substitutions without removing the `_1 = &_2` statement.
+///
+/// For simplicity, we treat raw pointers exactly like references.  This is sound because
+/// references impose strictly more constraints than raw pointers.
+pub struct PropagateReferences;
+
+impl<'tcx> MirPass<'tcx> for PropagateReferences {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        sess.mir_opt_level() >= 4
+    }
+
+    #[instrument(level = "trace", skip(self, tcx, body))]
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(def_id = ?body.source.def_id());
+        propagate_ssa(tcx, body);
+    }
+}
+
+fn propagate_ssa<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let mut ssa = SsaLocals::new(body);
+
+    let copy_classes = ssa.compute_copy_classes(body);
+    debug!(?copy_classes);
+
+    let mut fully_replacable_locals = fully_replacable_locals(&ssa, &copy_classes);
+    debug!(?fully_replacable_locals);
+
+    let targets = compute_local_values(tcx, body, &ssa, &fully_replacable_locals);
+    debug!(?targets);
+
+    let mut storage_to_remove = BitSet::new_empty(body.local_decls.len());
+    for (local, v) in targets.iter_enumerated() {
+        match v {
+            Value::Pointer(p) => {
+                if p.projection.first() == Some(&PlaceElem::Deref) {
+                    // We may introduce uses of a local after its `StorageDead`.
+                    // To avoid introducing UB, just remove the StorageDead statements.
+                    storage_to_remove.insert(p.local);
+                }
+            }
+            Value::Unknown => {
+                // We do not have a value, so it's definitely not replacable.
+                fully_replacable_locals.remove(local);
+            }
+        }
+    }
+
+    Replacer { tcx, targets, fully_replacable_locals, storage_to_remove }
+        .visit_body_preserves_cfg(body);
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum LocationExtended {
+    Plain(Location),
+    Arg,
+}
+
+#[derive(Debug)]
+struct SsaLocals {
+    /// Assignments to each local.  This defines whether the local is SSA.
+    assignments: IndexVec<Local, Set1<LocationExtended>>,
+    /// Number of "direct" uses of each local, ie. uses that are not dereferences.
+    direct_uses: IndexVec<Local, u32>,
+    /// We visit the body in reverse postorder, to ensure each local is assigned before it is used.
+    /// We remember the order in which we saw the assignments to compute the SSA values in a single
+    /// pass.
+    assignment_order: Vec<Local>,
+}
```

Comment:
This feels like something that other optimizations (dest prop?) would want to use, too. Is there any opportunity for code sharing here?

Reply:
We don't have other analyses based on SSA, but we definitely can. Probably when a use case arises?

Reply:
I would also be interested in this, because I think that SSA can be used to improve dataflow const prop (and the underlying framework). However, I'm not sure yet what exactly is required there, so I don't need this (yet).